### PR TITLE
Move datafusion-cli to new crate

### LINKED ISSUE #218: Move DataFusion cli/repl to own crate
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
The Datafusion cli/repl functionality now lives inside the DataFusion tree.
This has two main implications:

* Building DataFusion fetches extra dependencies (`clap` `rustyline` and transitive dependencies) unneeded for the rest of DataFusion, slowing the development cycle down a bit (mainly for clean builds).
* Installing the binary using cargo is a bit more cumbersome, `cargo install datafusion --bin datafusion-cli --release`
* dependency constraints on the cli dependencies potentially can cause conflicts for DataFusion consumers

**Describe the solution you'd like**
Move the `datafusion-cli` to the root dir.
Publish the crate to `crates.io` as `datafusion-cli`. The tool now can be installed using `cargo install datafusion-cli --release` and now lives outside of the DataFusion crate.

**Describe alternatives you've considered**
Keep structure as is.

**Additional context**

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #218

 # Rationale for this change
* Building DataFusion fetches extra dependencies (`clap` `rustyline` and transitive dependencies) unneeded for the rest of DataFusion, slowing the development cycle down a bit (mainly for clean builds).
* Installing the binary using cargo is currently a bit more cumbersome, `cargo install datafusion --bin datafusion-cli --release`
* dependency constraints on the cli dependencies potentially can cause conflicts for DataFusion consumers

| Version   |      Nr dependencies      |  
|----------|:-------------:|
| Master |  199 |
| PR |    186   |   

# What changes are included in this PR?

* Moves the source code & dependencies to datafusion-cli
* Drive-by change to make the default batch size a more reasonable value (8192 instead of >1M)
* Documentation updates
* Update `rustyline` version

# Are there any user-facing changes?

Users should be able to use datafusion-cli now instead of using the `bin` option inside of datafusion.

### CODE REVIEW

File: datafusion-cli/src/main.rs [hunk 1]
```diff
@@ -44,7 +44,7 @@ pub async fn main() {
         )
         .arg(
             Arg::with_name("batch-size")
-                .help("The batch size of each query, default value is 1048576")
```
Comment (andygrove) [inline datafusion-cli/src/main.rs:47]: We could get this value programmatically by creating a default config and getting this value.

Comment (Dandandan) [reply datafusion-cli/src/main.rs:47]: :+1: agreed

File: datafusion/docs/cli.md [hunk 2]
```diff
@@ -26,19 +26,19 @@ The DataFusion CLI is a command-line interactive SQL utility that allows queries
 Use the following commands to clone this repository and run the CLI. This will require the Rust toolchain to be installed. Rust can be installed from [https://rustup.rs/](https://rustup.rs/).
 
 ```sh
-git clone https://github.com/apache/arrow
-cd arrow/rust/datafusion
-cargo run --bin datafusion-cli --release
+git clone https://github.com/apache/arrow-datafusion
+cd arrow/datafusion-cli
```
Comment (andygrove) [inline datafusion/docs/cli.md:30]:
```suggestion
cd arrow-datafusion/datafusion-cli
```

File: datafusion/docs/cli.md [hunk 3]
```diff
@@ -26,19 +26,19 @@ The DataFusion CLI is a command-line interactive SQL utility that allows queries
...
 ```
 
 ## Run using Docker
 
 Use the following commands to clone this repository and build a Docker image containing the CLI tool. Note that there is `.dockerignore` file in the root of the repository that may need to be deleted in order for this to work.
 
 ```sh
-git clone https://github.com/apache/arrow
+git clone https://github.com/apache/arrow-datafusion
 cd arrow
```
Comment (andygrove) [inline datafusion/docs/cli.md:40]:
```suggestion
cd arrow-datafusion
```

**Comment by andygrove** [comment]
The Dockerfile will also need to be updated (it looks like it has incorrect paths). The Dockerfile also probably needs moving to the new datafusion-cli folder too.

File: datafusion-cli/src/main.rs (hunk 1 above)
Comment (Dandandan) [reply datafusion-cli/src/main.rs:47]: I made it inherit the DataFusion default now.

**Comment by Dandandan** [comment]
> The Dockerfile will also need to be updated (it looks like it has incorrect paths). The Dockerfile also probably needs moving to the new datafusion-cli folder too.

Thanks, I updated the Dockerfile

**Comment by andygrove** [comment]
LGTM. I can't approve this PR for some reason. Maybe a GitHub glitch .. I will check again later.

**Review by andygrove — APPROVED** [review]

**Comment by alamb** [comment]
LGTM too. Thanks @Dandandan

**Pull request merged**